var/spack/repos/builtin/packages/rust/package.py

```diff
@@ -4,59 +4,426 @@
+                            target=rust_arch['target']
+                        )
+                )
 
-    phases = ['configure', 'install']
+    # This routine returns the target architecture we intend to build for.
+    def get_rust_target(self):
+        if self.spec.satisfies('platform=linux target=x86_64:') or \
+           self.spec.satisfies('platform=cray target=x86_64:'):
+            return 'x86_64-unknown-linux-gnu'
+        elif self.spec.satisfies('platform=linux target=ppc64le:'):
+            return 'powerpc64le-unknown-linux-gnu'
+        elif self.spec.satisfies('platform=darwin target=x86_64:'):
+            return 'x86_64-apple-darwin'
+        else:
+            raise InstallError(
+                "rust-binary is not supported for '{0}'".format(
+                    self.spec.architecture
+                ))
 
     def configure(self, spec, prefix):
-        configure_args = [
-          '--prefix=%s' % prefix,
-          '--llvm-root=' + spec['llvm'].prefix,
-          # Workaround for "FileCheck does not exist" error
-          '--disable-codegen-tests',
-          # Includes Cargo in the build
-          # https://github.com/rust-lang/cargo/issues/3772#issuecomment-283109482
-          '--enable-extended',
-          # Prevent build from writing bash completion into system path
-          '--sysconfdir=%s' % join_path(prefix, 'etc/')
-          ]
-
-        configure(*configure_args)
-
-        # Build system defaults to searching in the same path as Spack's
-        # compiler wrappers which causes the build to fail
-        filter_file(
-            '#ar = "ar"',
-            'ar = "%s"' % join_path(spec['binutils'].prefix.bin, 'ar'),
-            'config.toml')
+        target = self.get_rust_target()
+        # See the NOTE above the resource loop - should be host architecture,
+        # not target aarchitecture if we're to support cross-compiling.
+        bootstrapping_install = Executable(
+            './spack_bootstrap_stage/rust-{version}-{target}/install.sh'
+            .format(
+                version=spec.version,
+                target=target
+            )
+        )
+        # install into the staging area
+        bootstrapping_install('--prefix={0}'.format(
+            join_path(self.stage.source_path, 'spack_bootstrap')
+        ))
+
+        boot_bin = join_path(self.stage.source_path, 'spack_bootstrap/bin')
+
+        # Always build rustc and cargo
+        tools = ['rustc', 'cargo']
+        # Only make additional components available in 'rust-bootstrap'
+        if '+rustfmt' in self.spec:
+            tools.append('rustfmt')
+        if '+analysis' in self.spec:
+            tools.append('analysis')
+        if '@1.33: +clippy' in self.spec:
+            tools.append('clippy')
+        if '+rls' in self.spec:
+            tools.append('rls')
+        if '+src' in self.spec:
+            tools.append('src')
+
+        ar = which('ar', required=True)
```

Comment:
I think the Intel compilers have their own ar called `xar`? Not sure if you care about supporting that.

Reply:
Yeah, it wasn't clear what the best way to deal with finding ar was. This _feels_ like something that other spack packages would have dealt with?

Reply:
Looks like it's actually "xiar". Isn't this perhaps something spack should take on? I'd prefer not to go down the path of testing alternative archivers...

Reply:
There are a few packages that do this manually (`cp2k`, `libint`, `libxc`, and `qt`). You don't have to bother with it if you don't want to, someone else can update the package if they really need it. I haven't heard any plans on modeling `ar` and `xiar` using Spack's compiler wrappers.

Reply:
Yeah, sorry, I'll leave it as an exercise somebody else, if needed.